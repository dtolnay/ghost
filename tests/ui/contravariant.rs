#![feature(generic_const_items)]
#![allow(incomplete_features)]

use ghost::phantom;

#[phantom]
struct ContravariantPhantom<#[contra] T>;

fn require_covariant<'a>(phantom: ContravariantPhantom<&'static str>) -> ContravariantPhantom<&'a str> {
    phantom
}

fn require_contravariant<'a>(phantom: ContravariantPhantom<&'a str>) -> ContravariantPhantom<&'static str> {
    phantom
}

fn main() {}
