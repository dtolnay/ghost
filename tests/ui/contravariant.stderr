error: lifetime may not live long enough
  --> tests/ui/contravariant.rs:10:5
   |
 9 | fn require_covariant<'a>(phantom: ContravariantPhantom<&'static str>) -> ContravariantPhantom<&'a str> {
   |                      -- lifetime `'a` defined here
10 |     phantom
   |     ^^^^^^^ returning this value requires that `'a` must outlive `'static`
